apps/builder/app/builder/features/style-panel/shared/css-value-input/css-value-input.stories.tsx

```diff
@@ -1,15 +1,91 @@
 import * as React from "react";
 import { Flex, DeprecatedTextField } from "@webstudio-is/design-system";
-import type { StyleValue } from "@webstudio-is/css-engine";
+import type {
+  InvalidValue,
+  StyleProperty,
+  StyleValue,
+} from "@webstudio-is/css-engine";
 import { CssValueInput, type IntermediateStyleValue } from "./css-value-input";
 import { action } from "@storybook/addon-actions";
 import { toValue } from "@webstudio-is/css-engine";
 import { theme } from "@webstudio-is/design-system";
+import { isValidDeclaration, parseCssValue } from "@webstudio-is/css-data";
 
 export default {
   component: CssValueInput,
 };
 
+export const WithCustomValidator = () => {
+  const [value, setValue] = React.useState<StyleValue>({
+    type: "unit",
+    value: 0,
+    unit: "px",
+  });
+
+  const [intermediateValue, setIntermediateValue] = React.useState<
+    StyleValue | IntermediateStyleValue
+  >();
+
+  return (
+    <CssValueInput
+      styleSource="preset"
+      property="boxShadow"
+      value={value}
+      intermediateValue={intermediateValue}
+      keywords={[]}
+      onChange={(value) => {
+        setIntermediateValue(value);
+      }}
+      onHighlight={(value) => {
+        action("onHighlight")(value);
+      }}
+      onChangeComplete={({ value }) => {
+        // on blur, select, enter etc.
+        setValue(value);
+        setIntermediateValue(undefined);
+        action("onChangeComplete")(value);
+      }}
+      onAbort={() => {
+        action("onAbort")();
+      }}
+      isValidDeclaration={(_, value) =>
+        isValidDeclaration("boxShadow", `${value} 0`)
+      }
+      parseIntermediateOrInvalidValue={(
+        _: StyleProperty,
+        styleValue: IntermediateStyleValue | InvalidValue
+      ): StyleValue => {
+        const value = styleValue.value.trim();
+        const unit = "unit" in styleValue ? styleValue.unit ?? "px" : "px";
+
+        /* Box shaodw needs atleast two values, so we are passing custom validator */
+        let styleInput = parseCssValue("boxShadow", `${value}${unit} 0`);
+        if (styleInput.type !== "invalid") {
+          return {
+            type: "unit",
+            value: parseFloat(value),
+            unit,
+          };
+        }
+
+        styleInput = parseCssValue("boxShadow", `${value} 0`);
+        if (styleInput.type !== "invalid") {
+          return {
+            type: "unit",
+            value: parseFloat(value),
+            unit,
+          };
+        }
+
+        return {
+          type: "invalid",
+          value,
+        };
```

Comment:
@istarkov @kof this is something i am just giving a try. What are your thoughts on this. The main reason is, `box-shadow` property doesn't work independently without multiple values. But the UI allows users to set them independently. Should there be a custom component, or re-use the existing one by passing custom validators something like this ??

Opened PR to test the theory. If sounds good, i can continue. Or else let's close and explore the ideas 👍